### PR TITLE
Add or, ior, and, iand, xor and ixor ops

### PR DESCRIPTION
**Pull request by aredridel**

### CODE REVIEW

File: lib/bn.js [hunk 1]
```diff
@@ -494,6 +494,100 @@ BN.prototype.neg = function neg() {
...
+
+// Or `num` with `this` in-place
+BN.prototype.ior = function ior(num) {
+  this.sign = this.sign || num.sign;
+
+  // a.length > b.length
+  var a;
+  var b;
+  if (this.length > num.length) {
+    a = this;
```
Comment (indutny) [inline lib/bn.js]: Does not seem like you are using `a` anywhere below.

File: lib/bn.js [hunk 2]
```diff
@@ -494,6 +494,100 @@ BN.prototype.neg = function neg() {
...
+    b = num;
+  } else {
+    a = num;
+    b = this;
+  }
+  for (var i = 0; i < b.length; i++) {
+    this.words[i] = this.words[i] | num.words[i];
```
Comment (indutny) [inline lib/bn.js]: Out of bounds reads.

Comment (indutny) [inline lib/bn.js]: Ah wait! I see now :)

Comment (indutny) [inline lib/bn.js]: While this loop does not produce out of bounds reads - it does not use higher limbs for `i` between `b.length` and `a.length`

File: lib/bn.js [hunk 3]
```diff
@@ -494,6 +494,100 @@ BN.prototype.neg = function neg() {
...
+
+// And `num` with `this` in-place
+BN.prototype.iand = function iand(num) {
+  this.sign = this.sign || num.sign;
+
+  // a.length > b.length
+  var a;
+  var b;
+  if (this.length > num.length) {
+    a = this;
```
Comment (indutny) [inline lib/bn.js]: `a` is not used.

File: lib/bn.js [hunk 4]
```diff
@@ -494,6 +494,100 @@ BN.prototype.neg = function neg() {
...
+    b = num;
+  } else {
+    a = num;
+    b = this;
+  }
+  for (var i = 0; i < b.length; i++) {
+    this.words[i] = this.words[i] & num.words[i];
```
Comment (indutny) [inline lib/bn.js]: Does not account for `i` between `b.length` and `a.length`.

File: test/bn-test.js [hunk 5]
```diff
@@ -445,4 +445,33 @@ describe('BN', function() {
     assert.equal(new BN(18).gcd(new BN(12)).toString(16), '6');
     assert.equal(new BN(-18).gcd(new BN(12)).toString(16), '6');
   });
+
+  it('should and numbers', function () {
+    assert.equal(new BN('1010101010101010101010101010101010101010', 2)
```
Comment (indutny) [inline test/bn-test.js]: 80 column limit please.

**Comment by indutny** [comment]
Looks awesome, thank you! A couple of nits and we are good to go ;)

**Comment by aredridel** [comment]
Tweaking those now!

Also may be adding a `setn(bit, value)` since setting a single bit seems to be missing.

**Comment by aredridel** [comment]
Updated to address your comments.

File: lib/bn.js [hunk 6]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+  }
+
+  for (; i < a.length; i++) {
```
Comment (indutny) [inline lib/bn.js]: I guess this is required only if `a === this`?

File: lib/bn.js [hunk 7]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+    this.words[i] = 0;
+  }
+  this.length = b.length;
+
+  return this._normSign();
```
Comment (indutny) [inline lib/bn.js]: Let's use `.strip()` here.

File: lib/bn.js [hunk 8]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+    this.words[i] = a.words[i] ^ b.words[i];
+  }
+
+  for (; i < a.length; i++) {
+    this.words[i] = a.words[i] ^ 0;
+  }
+
+  this.length = a.length;
+
+  return this._normSign();
```
Comment (indutny) [inline lib/bn.js]: `this.strip()`, please

File: lib/bn.js [hunk 9]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+  else
+    return num.clone().ixor(this);
+};
+
+
+// Set `bit` of `this`
+BN.prototype.setn = function setn(bit, val) {
+  assert(typeof bit === 'number' && bit >= 0);
+
+  var off = Math.floor(bit / 26);
```
Comment (indutny) [inline lib/bn.js]: `(bit / 26) | 0;`

File: lib/bn.js [hunk 10]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+  var wbit = bit % 26;
+  while (this.length <= off) {
+      this.words[this.length++] = 0;
+  }
+
+  if (val) {
+      this.words[off] = this.words[off] | (1 << wbit);
```
Comment (indutny) [inline lib/bn.js]: Sorry, I'm using 2-space indent everywhere.

File: lib/bn.js [hunk 11]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+};
+
+
+// Set `bit` of `this`
+BN.prototype.setn = function setn(bit, val) {
+  assert(typeof bit === 'number' && bit >= 0);
+
+  var off = Math.floor(bit / 26);
+  var wbit = bit % 26;
+  while (this.length <= off) {
```
Comment (indutny) [inline lib/bn.js]: Please omit braces when block has only one line.

File: test/bn-test.js [hunk 12]
```diff
@@ -445,4 +445,60 @@ describe('BN', function() {
...
+    assert.equal(new BN('11001100110011001100110011001100', 2)
+                 .ixor(new BN('1', 2))
+                 .toString(2), '11001100110011001100110011001101');
+    assert.equal(new BN('1', 2)
+                 .ixor(new BN('11001100110011001100110011001100', 2))
+                 .toString(2), '11001100110011001100110011001101');
+  });
+
+  it('should allow single bits to be set', function () {
+      assert.equal(new BN(0).setn(2, true).toString(2), '100');
```
Comment (indutny) [inline test/bn-test.js]: Two-space indent, please.

**Comment by indutny** [comment]
Yay, even better now! Thank you.

Getting very close to merging it, please (hopefully) last series of nits.

**Comment by aredridel** [comment]
Does sign handling seem right?

File: lib/bn.js [hunk 13]
```diff
@@ -494,6 +494,124 @@ BN.prototype.neg = function neg() {
...
+  if (this.length > num.length)
+    return this.clone().ior(num);
+  else
+    return num.clone().ior(this);
+};
+
+
+// And `num` with `this` in-place
+BN.prototype.iand = function iand(num) {
+  this.sign = this.sign || num.sign;
```
Comment (indutny) [inline lib/bn.js]: I'd use `&&` here. This is how EcmaScript handles it.

File: lib/bn.js (hunk 6 above)
Comment (aredridel) [inline lib/bn.js]: Yes.

**Comment by aredridel** [comment]
Another update!

File: lib/bn.js (hunk 6 above)
Comment (indutny) [inline lib/bn.js]: Actually there is no point in this, please just set the length below as you do and remove the loop.

File: lib/bn.js [hunk 14]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
   return r;
 };
 
+
+// Or `num` with `this` in-place
+BN.prototype.ior = function ior(num) {
+  this.sign = this.sign || num.sign;
+
+  while (this.length < num.length) this.words[this.length++] = 0;
```
Comment (indutny) [inline lib/bn.js]:
```javascript
while (...)
  statement;
```

File: lib/bn.js [hunk 15]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+
+  for (var i = 0; i < num.length; i++) {
```
Comment (indutny) [inline lib/bn.js]:
```javascript
for (...)
  statement;
```

File: lib/bn.js [hunk 16]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+  var a;
+  var b;
+  if (this.length > num.length) {
+    a = this;
+    b = num;
+  } else {
+    a = num;
+    b = this;
+  }
+  for (var i = 0; i < b.length; i++) {
```
Comment (indutny) [inline lib/bn.js]:
```javascript
for (...)
  statement;
```

File: lib/bn.js [hunk 17]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+    a = num;
+    b = this;
+  }
+
+  for (var i = 0; i < b.length; i++) {
+    this.words[i] = a.words[i] ^ b.words[i];
+  }
+
+  for (; i < a.length; i++) {
+    this.words[i] = a.words[i] ^ 0;
```
Comment (indutny) [inline lib/bn.js]: Just `a.words[i]`, may be?

File: lib/bn.js [hunk 18]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+  var b;
+  if (this.length > num.length) {
+    a = this;
+    b = num;
+  } else {
+    a = num;
+    b = this;
+  }
+
+  for (var i = 0; i < b.length; i++) {
```
Comment (indutny) [inline lib/bn.js]:
```javascript
for (...)
  statement;
```

File: lib/bn.js [hunk 19]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+    this.words[i] = a.words[i] ^ b.words[i];
+  }
+
+  for (; i < a.length; i++) {
```
Comment (indutny) [inline lib/bn.js]:
```javascript
for (...)
  statement;
```

File: lib/bn.js [hunk 20]
```diff
@@ -494,6 +494,123 @@ BN.prototype.neg = function neg() {
...
+// Set `bit` of `this`
+BN.prototype.setn = function setn(bit, val) {
+  assert(typeof bit === 'number' && bit >= 0);
+
+  var off = (bit / 26) | 0;
+  var wbit = bit % 26;
+
+  while (this.length <= off) this.words[this.length++] = 0;
+
+  if (val) {
```
Comment (indutny) [inline lib/bn.js]:
```javascript
if (...)
  stmt;
else
  stmt;
```

File: test/bn-test.js [hunk 21]
```diff
@@ -445,4 +445,60 @@ describe('BN', function() {
...
+                 .ixor(new BN('1', 2))
+                 .toString(2), '11001100110011001100110011001101');
+    assert.equal(new BN('1', 2)
+                 .ixor(new BN('11001100110011001100110011001100', 2))
+                 .toString(2), '11001100110011001100110011001101');
+  });
+
+  it('should allow single bits to be set', function () {
+    assert.equal(new BN(0).setn(2, true).toString(2), '100');
+    assert.equal(new BN(0).setn(27, true).toString(2), '1000000000000000000000000000');
```
Comment (indutny) [inline test/bn-test.js]: 80 column! :)

**Comment by indutny** [comment]
@aredridel thanks again! And again another series of nits, sorry for this!

**Comment by aredridel** [comment]
Not a problem! It's a precise style, I don't mind matching. It's just alien to krakenjs style.

**Comment by aredridel** [comment]
There. Got it all, I think.

File: lib/bn.js (hunk 1 above)
Comment (dcousens) [inline lib/bn.js]: `a` is still unused @aredridel, incase you missed this nit :)

File: lib/bn.js [hunk 22]
```diff
@@ -494,6 +494,118 @@ BN.prototype.neg = function neg() {
...
+    b = num;
+  } else {
+    a = num;
+    b = this;
+  }
+
+  for (var i = 0; i < b.length; i++)
+    this.words[i] = a.words[i] ^ b.words[i];
+
+  for (; i < a.length; i++)
```
Comment (indutny) [inline lib/bn.js]: Please put `if (this !== a)` on a previous line ;)

File: lib/bn.js (hunk 1 above)
Comment (indutny) [inline lib/bn.js]: Yeah.

**Comment by indutny** [comment]
Two last nits, otherwise LGTM!

**Comment by aredridel** [comment]
Only saw one further comment with a nit, but fixed!

**Comment by indutny** [comment]
Thank you! Landed and released in 1.3.0

**Comment by aredridel** [comment]
Awesome!

**Pull request closed without merging**